experiments/process_sandbox/README.md

```diff
@@ -330,3 +329,184 @@ Code layout
+  |                                                                                                                                          
+  |                                                                                                                                            |-- 0x7fbe80000000 [r--s]
+  |                                                                                                                                            |
+  ...                                                                                           |-- 0x7fbf00000000 [r--s]                      |
+  |                                                                                             |                                              |
+  |                                              |-- 0x7fbf80000000 [r--s]                      |                                              |
+  |                                              |                                              |                                              |        PageMap
+  |-- 0x7fc000000000                             |                                              |                                              |
+                                                 |                                              |        PageMap                               |
+                                                 |                                              |                                              |
+                                                 |        PageMap                               |                                              | 
+  |-- 0x7ffe80000000 -- 0x7ffec0000000 [rw-s]    |                                              |                                              |-- child 3 region [rw-s]
+                                                 |                                              |                                            
+  |-- 0x7fff00000000 -- 0x7fff40000000 [rw-s]    |                                              |-- child 2 region [rw-s]                     
+                                                 |                                                                                           
+  |-- 0x7fff80000000 -- 0x7fffc0000000 [rw-s]    |-- child 1 region [rw-s]                                                                   
+
+```
+
+The `library_runner` is responsible, at initialization time, for ensuring that both the `PageMap` and the child's memory region (i.e., heap) are correctly mapped in the process's address space.
+For that purposes, it uses the `memfd` file descriptors inherited from the parent.
+It relies on `mmap` to install the heap as a shared memory region that is both readable and writable.
+For the `PageMap`, it resizes the region and drops write access to the corresponding file descriptor using `mmap`.
```

Comment:
Does it?  It shouldn't resize the region at all, and the parent should be dropping write access when it sets up the sandbox policy.

On platforms with well-designed security mechanisms, [it doesn't have the rights to do anything other than a read-only mapping of the shared-memory object](https://github.com/microsoft/verona/blob/b5be27ea4730301bc893d8746fa758741b2e0ab4/experiments/process_sandbox/include/process_sandbox/platform/sandbox_capsicum.h#L77).  On Linux, [it can't map the memory read-write](https://github.com/microsoft/verona/blob/b5be27ea4730301bc893d8746fa758741b2e0ab4/experiments/process_sandbox/include/process_sandbox/platform/sandbox_seccomp-bpf.h#L52) (and there's a [test that checks that it can't](https://github.com/microsoft/verona/blob/b5be27ea4730301bc893d8746fa758741b2e0ab4/experiments/process_sandbox/tests/sandboxlib-modify-pagemap.cc#L22)).  It can resize the file on Linux and so we close the file descriptor before we run any untrusted code.  We should fix this using the sealing mechanism on memfd, but we'd still need to close the file descriptor because we fall back to POSIX shared memory if `memfd` doesn't work.

Reply:
I will clarify that part of the text, I mixed things up.
For the `PageMap`, as you pointed out, we get the following trace in the child:

```
openat(AT_FDCWD, "/proc/self/fd/4", O_RDONLY) = 8
dup2(8, 4)                              = 4
close(8)                                = 0
```

And later on, we map the content with:

```
mmap(NULL, 274877906944, PROT_READ, MAP_SHARED, 4, 0) = 0x7fbff7a1a000
```

Reformulation: 

```
The `library_runner` opens the `memfd` file descriptor that corresponds to the `PageMap` object and maps its content read-only in the child's address space.
```